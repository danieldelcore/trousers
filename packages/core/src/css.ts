import hash from '@trousers/hash';

import { Definition, CSSObject } from './types';
import themify from './themify';

export interface CollectorReturn {
    _get: () => Definition[];
    modifier: (id: string, styles: CSSObject) => CollectorReturn;
    global: (styles: CSSObject) => CollectorReturn;
    theme: (theme: Record<string, any>) => CollectorReturn;
}

function css(styles: CSSObject): CollectorReturn;
function css(
    idOrStyle: string | CSSObject,
    styles?: CSSObject,
): CollectorReturn;
function css(
    idOrStyle: string | CSSObject,
    styles?: CSSObject,
): CollectorReturn {
    let id = typeof idOrStyle === 'string' ? idOrStyle : '';
    let styleObject = (typeof idOrStyle === 'string'
        ? styles
        : idOrStyle) as CSSObject;

    const elementId = `${id}-${hash(JSON.stringify(styleObject))}`;
    const styleMap: Definition[] = [
        {
            id,
            type: 'element',
            className: `.${elementId}`,
            styles: styleObject,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                id: modifierId,
                type: 'modifier',
                className: `.${elementId}--${modifierId}-${hash(
                    JSON.stringify(modifierStyles),
                )}`,
                styles: modifierStyles,
            });

            return self;
        },
        global: globalStyles => {
            styleMap.push({
                id: `global-${id}-${hash(JSON.stringify(globalStyles))}`,
                type: 'global',
                className: '',
                styles: globalStyles,
            });

            return self;
        },
        theme: theme => {
            const themeHash = hash(JSON.stringify(theme));
            styleMap.push({
                id: `theme-${id}-${themeHash}`,
                type: 'theme',
                className: `.theme-${id}-${themeHash}`,
                styles: themify(theme),
            });

            return self;
        },
    };

    return self;
}

export type Collector = typeof css;

export default css;
