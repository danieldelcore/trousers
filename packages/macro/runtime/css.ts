interface Definition {
    type: 'element' | 'modifier' | 'global' | 'theme';
    id: string;
    styles: string;
}

interface CollectorReturn {
    modifier: (id: string, styles: string) => CollectorReturn;
    global: (id: string, styles: string) => CollectorReturn;
    theme: (id: string, styles: string) => CollectorReturn;
    _get: () => Definition[];
}

function css(elementId: string, elementStyles: string) {
    const styleMap: Definition[] = [
        {
            type: 'element',
            id: elementId,
            styles: elementStyles,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                type: 'modifier',
                id: modifierId,
                styles: modifierStyles,
            });

            return self;
        },
        global: (globalId, globalStyles) => {
            styleMap.push({
                type: 'global',
                id: globalId,
                styles: globalStyles,
            });

            return self;
        },
        theme: (themeId, themeStyles) => {
            styleMap.push({
                type: 'theme',
                id: themeId,
                styles: themeStyles,
            });

            return self;
        },
    };

    return self;
}

export default css;
