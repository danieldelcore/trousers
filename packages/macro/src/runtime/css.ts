interface Definition {
    type: 'element' | 'modifier' | 'global' | 'theme';
    id: string;
    styles: Record<string, string>;
}

export interface CollectorReturn {
    modifier: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    global: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    theme: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    _get: () => Definition[];
}

function css(elementId: Definition['id'], elementStyles: Definition['styles']) {
    const styleMap: Definition[] = [
        {
            type: 'element',
            id: elementId,
            styles: elementStyles,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                type: 'modifier',
                id: modifierId,
                styles: modifierStyles,
            });

            return self;
        },
        global: (globalId, globalStyles) => {
            styleMap.push({
                type: 'global',
                id: globalId,
                styles: globalStyles,
            });

            return self;
        },
        theme: (themeId, themeStyles) => {
            styleMap.push({
                type: 'theme',
                id: themeId,
                styles: themeStyles,
            });

            return self;
        },
    };

    return self;
}

export default css;
