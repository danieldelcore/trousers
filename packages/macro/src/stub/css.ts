/* eslint-disable @typescript-eslint/no-unused-vars */
import {
    CSSObject,
    Collector as CoreCollector,
    CollectorReturn as CoreCollectorReturn,
} from '@trousers/core';

export type CollectorReturn = CoreCollectorReturn;
export type Collector = CoreCollector;

function css(_styles: CSSObject): CollectorReturn;
function css(
    _idOrStyle: string | CSSObject,
    _styles?: CSSObject,
): CollectorReturn;
function css(
    _idOrStyle: string | CSSObject,
    _styles?: CSSObject,
): CollectorReturn {
    const self: CollectorReturn = {
        modifier: (_modifierId, _modifierStyles) => self,
        global: _globalStyles => self,
        theme: _theme => self,
        _get: () => [],
    };

    return self;
}

export default css;
