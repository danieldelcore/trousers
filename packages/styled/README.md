# @trousers/styled

`@trousers/styles` is an implementation of Trousers for lovers of the Styled Components API (with a few tweaks).

## Installation

`yarn add @trousers/styled` or `npm install @trousers/styled`

## API

### `styled`

The `styled` method is a factory function which provides you with a way to quickly create and style components.
It is able to be used in two separate ways...

1. Dot notation `styled.div`
2. Function notation `styled('div')`

**Arguments:**

-   `tagId` (`string`)?: The id of the element tag you wish to create (ie. div, button, span, etc)

**Returns:**

-   `Record<Elements>`: HTML elements ids as an object (see Dot notation)
-   `(styles: Collector) => ReactComponent`: Returns a method which accepts a css collector

**Example:**

```js
import styled, { css } from '@trousers/styled';

// Dot notation
const StyledDiv = styled.div(css('element', { color: 'red' }));

// Function notation
const StyledDiv = styled('div')(css('element', { color: 'red' }));
```
