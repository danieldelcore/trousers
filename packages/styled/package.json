{
    "name": "@trousers/styled",
    "version": "4.0.0-alpha3",
    "main": "dist/trousers-styled.cjs.js",
    "module": "dist/trousers-styled.esm.js",
    "types": "dist/index.d.ts",
    "license": "MIT",
    "dependencies": {
        "@trousers/core": "4.0.0-alpha3",
        "@trousers/react": "4.0.0-alpha3"
    },
    "devDependencies": {
        "react": "^17.0.1"
    },
    "peerDependencies": {
        "react": ">=16.8.0"
    },
    "repository": "https://github.com/danieldelcore/trousers/tree/master/packages/styled"
}
