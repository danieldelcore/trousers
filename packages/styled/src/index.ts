export { default } from './styled';
export { default as css } from '@trousers/core';
