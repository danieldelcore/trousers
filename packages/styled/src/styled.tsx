/** @jsx jsx */
import { ComponentType, forwardRef } from 'react';
import jsx from '@trousers/react';
import { Collector } from '@trousers/core';

type Elements = keyof JSX.IntrinsicElements;
const domElements: Elements[] = [
    'a',
    'abbr',
    'address',
    'area',
    'article',
    'aside',
    'audio',
    'b',
    'base',
    'bdi',
    'bdo',
    'big',
    'blockquote',
    'body',
    'br',
    'button',
    'canvas',
    'caption',
    'cite',
    'code',
    'col',
    'colgroup',
    'data',
    'datalist',
    'dd',
    'del',
    'details',
    'dfn',
    'dialog',
    'div',
    'dl',
    'dt',
    'em',
    'embed',
    'fieldset',
    'figcaption',
    'figure',
    'footer',
    'form',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'head',
    'header',
    'hgroup',
    'hr',
    'html',
    'i',
    'iframe',
    'img',
    'input',
    'ins',
    'kbd',
    'keygen',
    'label',
    'legend',
    'li',
    'link',
    'main',
    'map',
    'mark',
    'menu',
    'menuitem',
    'meta',
    'meter',
    'nav',
    'noscript',
    'object',
    'ol',
    'optgroup',
    'option',
    'output',
    'p',
    'param',
    'picture',
    'pre',
    'progress',
    'q',
    'rp',
    'rt',
    'ruby',
    's',
    'samp',
    'script',
    'section',
    'select',
    'small',
    'source',
    'span',
    'strong',
    'style',
    'sub',
    'summary',
    'sup',
    'table',
    'tbody',
    'td',
    'textarea',
    'tfoot',
    'th',
    'thead',
    'time',
    'title',
    'tr',
    'track',
    'u',
    'ul',
    'var',
    'video',
    'wbr',
    'circle',
    'clipPath',
    'defs',
    'ellipse',
    'foreignObject',
    'g',
    'image',
    'line',
    'linearGradient',
    'marker',
    'mask',
    'path',
    'pattern',
    'polygon',
    'polyline',
    'radialGradient',
    'rect',
    'stop',
    'svg',
    'text',
    'tspan',
];

interface StyledProps extends JSX.IntrinsicAttributes {
    // TODO: replace below with template literal type for $primary
    [key: string]: any;
}
type CollectorHOC = (css: ReturnType<Collector>) => ComponentType<StyledProps>;
type ElementMap = {
    [key in keyof JSX.IntrinsicElements]?: CollectorHOC;
};

interface Styled extends ElementMap {
    (Tag: keyof JSX.IntrinsicElements): CollectorHOC;
}

let styledBase = (Tag: keyof JSX.IntrinsicElements) => (
    css: ReturnType<Collector>,
) =>
    forwardRef<HTMLElement, StyledProps>((props, ref) => {
        //@ts-ignore union type is too complex
        return <Tag css={css} {...props} ref={ref} />;
    });

const styled: Styled = Object.assign(
    styledBase,
    domElements.reduce<ElementMap>((accum, element) => {
        accum[element] = styledBase(element);
        return accum;
    }, {}),
);

export default styled;
