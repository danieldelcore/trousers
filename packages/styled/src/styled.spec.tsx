import React from 'react';
import css from '@trousers/core';
import styled from './styled';
import { render } from '@testing-library/react';

describe('styled', () => {
    it('creates a styled component from dot notation element id', () => {
        const StyledDiv = styled.div(
            css('element', {
                color: 'red',
            }),
        );

        const { container } = render(<StyledDiv />);

        expect(container).toMatchSnapshot();
    });

    it('creates a styled component from string element id', () => {
        const StyledDiv = styled('div')(
            css('element', {
                color: 'red',
            }),
        );

        const { container } = render(<StyledDiv />);

        expect(container).toMatchSnapshot();
    });

    it('passes through html attributes', () => {
        const StyledDiv = styled('div')(
            css('element', {
                color: 'red',
            }),
        );

        const { container } = render(<StyledDiv data-testid="foo" />);

        expect(container).toMatchSnapshot();
    });

    it('passes through modifier props', () => {
        const StyledDiv = styled('div')(
            css('element', {
                color: 'red',
            }).modifier('primary', { color: 'green' }),
        );

        const { container } = render(<StyledDiv $primary />);

        expect(container).toMatchSnapshot();
    });

    it('passes through multiple modifier props', () => {
        const StyledDiv = styled('div')(
            css('element', {
                color: 'red',
            })
                .modifier('primary', { color: 'green' })
                .modifier('secondary', { color: 'purple' }),
        );

        const { container } = render(<StyledDiv $primary $secondary />);

        expect(container).toMatchSnapshot();
    });
});
