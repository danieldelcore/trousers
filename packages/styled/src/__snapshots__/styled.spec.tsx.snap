// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled creates a styled component from dot notation element id 1`] = `
<div>
  <div
    class="element-2313942302"
  />
</div>
`;

exports[`styled creates a styled component from string element id 1`] = `
<div>
  <div
    class="element-2313942302"
  />
</div>
`;

exports[`styled passes through html attributes 1`] = `
<div>
  <div
    class="element-2313942302"
    data-testid="foo"
  />
</div>
`;

exports[`styled passes through modifier props 1`] = `
<div>
  <div
    class="element-2313942302 element-2313942302--primary-2402939536"
  />
</div>
`;

exports[`styled passes through multiple modifier props 1`] = `
<div>
  <div
    class="element-2313942302 element-2313942302--primary-2402939536 element-2313942302--secondary-3026956261"
  />
</div>
`;
